### ISSUE TITLE
Issues when building for SVE

### ISSUE THREAD

**Issue by adegomme** [arbor-sim/arbor#2101]
version : all > 0.6

**Describe the bug**

When building (or cross building) with SVE enabled (for benchmarking purposes), we encountered a few issues in the generated portions of code, with various compilers (arm/gcc/llvm).

Process is to build modcc for the native system and use it to build the SVE version, as documentation says, so:

```
mkdir -p $BUILDDIR
cd $BUILDDIR
git clone --depth=1 --recursive https://github.com/arbor-sim/arbor.git arbor
cd $BUILDDIR/arbor
mkdir build_modcc
cd build_modcc
cmake ..
make -j modcc 
mkdir -p $BUILDDIR/arbor/build
cd $BUILDDIR/arbor/build
cmake -DCMAKE_INSTALL_PREFIX=/usr/local/arbor -D ARB_WITH_MPI=ON -D ARB_WITH_PYTHON=ON -D ARB_VECTORIZE=ON -D CMAKE_CXX_FLAGS="-O3 -mcpu=neoverse-v1" -D CMAKE_C_FLAGS="-O3 -mcpu=neoverse-v1 " -D ARB_MODCC=$BUILDDIR/arbor/build_modcc/bin/modcc -D ARB_ARCH=armv8.4-a+sve $BUILDDIR/arbor 
cmake --build $BUILDDIR/arbor/build --target all -- -j$(nproc) 
```

The first issue encountered was

```
/var/tmp/arbor/build/mechanisms/generated/bbp/K_Tst_cpu.cpp:37:27: error: constexpr variable 'min_align_' must be initialized by a constant expression
static constexpr unsigned min_align_ = std::max(S::min_align(simd_value{}), S::min_align(simd_index{}));
/var/tmp/arbor/build/mechanisms/generated/bbp/K_Tst_cpu.cpp:37:72: note: non-literal type 'arb::bbp_catalogue::kernel_K_Tst::simd_value' (aka '__SVFloat64_t') cannot be used in a constant expression
static constexpr unsigned min_align_ = std::max(S::min_align(simd_value{}), S::min_align(simd_index{}));
```

Removing the "constexpr" in the cprinter is enough to go past it, but I don't know if performance may be affected here.

`sed -i 's/static constexpr unsigned min_align_/static unsigned min_align_/g' modcc/printer/cprinter.cpp`

The second issue is

```
/var/tmp/arbor/build/mechanisms/generated/default/inject_cpu.cpp:130:14: error: invalid operands of types 'arb::default_catalogue::kernel_inject::simd_value' {aka 'svfloat64_t'} and 'arb::default_catalogue::kernel_inject::simd_value' {aka 'svfloat64_t'} to binary 'operator-'
  130 |           xd -= t_xd0_;
```

I was able to compile by using "S::sub" and not -= , but I'm not sure about the correctness of the computation, which is why I'n not pushing a PR. I also have no idea about other SIMD behaviour here.

`
sed -i -e 's/{} -= t_{}0_/{} = S::sub({}, t_{}0_)/' -e 's/scaled, nam/scaled, scaled, nam/' modcc/printer/cprinter.cpp 
`

**Context**
OS: linux,
Arbor version: trunk, but all > 0.6 ( since 4a305b4f946ee857245f674f7b1d9e56929aaf84 which introduced the -= lines
Arch: aarch64 + sve

**Comment by boeschf** [comment]
@adegomme thanks for letting us know. Can you check whether the following branch works for you?
https://github.com/boeschf/arbor/tree/fix/sve_bugs

**Comment by adegomme** [comment]
I can confirm this compiles (gcc, armclang and llvm).

To make it pass most of the testsuite on v1, I had to change a small value in simd.hpp, as the default size of the vector was assumed to be 512, which was then breaking asserts when running tests, as v1 uses 256.

sed -i "s/default_width = 8/default_width = 4/" modcc/printer/simd.hpp

(this is not a good solution, only a workaround for my case).
With this, only 3 simd tests fail (those testing vectors of size > 256, so that's expected).

Sadly I fail to find representative benchmarks/inputs where SVE is used with arbor, it looks like most of the ones I tested (bench, or nsuite/ring) are not using it ? are there some examples you could advise, maybe?

**Comment by boeschf** [comment]
I can't come up with a nice solution to this problem, so I propose an additional configure time switch `ARB_SVE_WIDTH` and have updated the branch accordingly. If you set this to `4`, it should do what you proposed above and disable the larger tests. Please have a look and let me know.

Regarding benchmarks, we haven't done many benchmarks recently on ARM (with SVE enabled). The most recent run was on a single Graviton 3E using the busyring benchmark, but the results posted used NEON, https://arbor-sim.org/attachments/2023-03-27-hbp-summit-satellite-event_FB.pdf, and it was using an earlier version of master without the fixed-dt feature that we merged recently. I can find the exact run times and configs if desired.

**Comment by adegomme** [comment]
I was attempting to confirm everything was working, but I notice that the last commit was not enough and brought back the constexpr issue, apparently

`/var/tmp/arbor/build/mechanisms/generated/allen/Im_cpu.cpp:37:27: error: constexpr variable 'min_align_' must be initialized by a constant expression
static constexpr unsigned min_align_ = std::max(S::min_align(simd_value{}), S::min_align(simd_index{}));
                          ^            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/arbor/build/mechanisms/generated/allen/Im_cpu.cpp:37:72: note: non-literal type 'simd_value' (aka '__SVFloat64_t') cannot be used in a constant expre                  ssion
static constexpr unsigned min_align_ = std::max(S::min_align(simd_value{}), S::min_align(simd_index{}));
`

**Comment by adegomme** [comment]
There is (without the constexpr) one test remaining that fails

```
[ RUN      ] S/sizeless_api/1.arithmetic
/var/tmp/arbor/test/unit/test_simd.cpp:1744: Failure
Value of: testing::indexed_almost_eq_n(N, c, expected)
  Actual: false (floating point numbers 1 and nan differ at index 2)
Expected: true
[  FAILED  ] S/sizeless_api/1.arithmetic, where TypeParam = simd_types_t<simd_t<__SVFloat64_t, double, 4u>, simd_t<__SVInt64_t, int, 4u>, simd_t<__SVBool_t, bool, 4u> > (0 ms)


[==========] 885 tests from 165 test suites ran. (6626 ms total)
[  PASSED  ] 884 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] S/sizeless_api/1.arithmetic, where TypeParam = simd_types_t<simd_t<__SVFloat64_t, double, 4u>, simd_t<__SVInt64_t, int, 4u>, simd_t<__SVBool_t, bool, 4u> >

 1 FAILED TEST
```

l1744 is pow, no idea what can be the issue.

**Comment by boeschf** [comment]
I go ahead and close this issue, as it should be fixed with #2148.
